File: problems/380/380.insert-delete-get-random-o-1.py
#
# @lc app=leetcode id=380 lang=python3
#
# [380] Insert Delete GetRandom O(1)
#

# @lc code=start
import random
from collections import defaultdict


class RandomizedSet:
    # keep track of values in a set
    # get random value from set
    # average O(1) time
    # set keeps track of index
    # store list of vals for random

    def __init__(self):
        self.set_vals = defaultdict(int)  # holds index of value in list_vals
        self.list_vals = []

    def insert(self, val: int) -> bool:
        if val in self.set_vals:
            return False

        self.list_vals.append(val)
        self.set_vals[val] = len(self.list_vals) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.set_vals:
            return False

        # if last element, just pop
        if self.set_vals[val] == len(self.list_vals) - 1:
            self.set_vals.pop(val)
            self.list_vals.pop()
        # if not last element, swap with last and pop
        else:
            idx = self.set_vals[val]
            last_val = self.list_vals[-1]

            self.list_vals[idx] = last_val
            self.set_vals[last_val] = idx

            self.set_vals.pop(val)
            self.list_vals.pop()

        return True

    def getRandom(self) -> int:
        rand_i = random.randint(0, len(self.list_vals) - 1)
        return self.list_vals[rand_i]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
# @lc code=end

File: problems/380/380.insert-delete-get-random-o-1.go
/*
 * @lc app=leetcode id=380 lang=golang
 *
 * [380] Insert Delete GetRandom O(1)
 */

package solution

// @lc code=start
import "math/rand"

type RandomizedSet struct {
	vals map[int]int
	data []int
}

func Constructor() RandomizedSet {
	vals := make(map[int]int)
	data := make([]int, 0)
	return RandomizedSet{vals, data}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.vals[val]; ok {
		return false
	}

	this.vals[val] = len(this.data)
	this.data = append(this.data, val)
	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	if _, ok := this.vals[val]; !ok {
		return false
	}

	last_ele_idx := len(this.data) - 1
	last_ele := this.data[last_ele_idx]
	curr_ele_idx := this.vals[val]

	// swap idx
	this.vals[last_ele] = curr_ele_idx
	this.data[curr_ele_idx], this.data[last_ele_idx] = this.data[last_ele_idx], this.data[curr_ele_idx]

	// remove
	delete(this.vals, val)
	this.data = this.data[:last_ele_idx]

	return true
}

func (this *RandomizedSet) GetRandom() int {
	rand_idx := rand.Intn(len(this.data))
	return this.data[rand_idx]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */
// @lc code=end
